iothub/device/samples/convention-based-samples/Thermostat/ThermostatSample.cs

```diff
@@ -0,0 +1,205 @@
+using Microsoft.Extensions.Logging;
+using Newtonsoft.Json;
+
+namespace Microsoft.Azure.Devices.Client.Samples
+{
+    public class ThermostatSample
+    {
+        private readonly Random _random = new Random();
+
+        private double _temperature = 0d;
+        private double _maxTemp = 0d;
+
+        // Dictionary to hold the temperature updates sent over.
+        // NOTE: Memory constrained devices should leverage storage capabilities of an external service to store this information and perform computation.
+        // See https://docs.microsoft.com/en-us/azure/event-grid/compare-messaging-services for more details.
+        private readonly Dictionary<DateTimeOffset, double> _temperatureReadingsDateTimeOffset = new Dictionary<DateTimeOffset, double>();
+
+        private readonly DeviceClient _deviceClient;
+        private readonly ILogger _logger;
+
+        public ThermostatSample(DeviceClient deviceClient, ILogger logger)
+        {
+            _deviceClient = deviceClient ?? throw new ArgumentNullException($"{nameof(deviceClient)} cannot be null.");
+            _logger = logger ?? LoggerFactory.Create(builer => builer.AddConsole()).CreateLogger<ThermostatSample>();
+        }
+
+        public async Task PerformOperationsAsync(CancellationToken cancellationToken)
+        {
+            // This sample follows the following workflow:
+            // -> Set handler to receive and respond to writable property update requests.
+            // -> Set handler to receive and respond to commands.
+            // -> Periodically send "temperature" over telemetry.
+            // -> Send "maxTempSinceLastReboot" over property update, when a new max temperature is set.
+
+            _logger.LogDebug($"Subscribe to writable property updates.");
+            await _deviceClient.SubscribeToWritablePropertiesEventAsync(HandlePropertyUpdatesAsync, null, cancellationToken);
+
+            _logger.LogDebug($"Subscribe to commands.");
+            await _deviceClient.SubscribeToCommandsAsync(HandleCommandsAsync, null, cancellationToken);
+
+            bool temperatureReset = true;
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                if (temperatureReset)
+                {
+                    // Generate a random value between 5.0°C and 45.0°C for the current temperature reading.
+                    _temperature = Math.Round(_random.NextDouble() * 40.0 + 5.0, 1);
+                    temperatureReset = false;
+                }
+
+                await SendTemperatureAsync();
+                await Task.Delay(5 * 1000);
+            }
+        }
+
+        // The callback to handle property update requests.
+        private async Task HandlePropertyUpdatesAsync(ClientPropertyCollection writableProperties, object userContext)
+        {
+            foreach (KeyValuePair<string, object> writableProperty in writableProperties)
+            {
+                switch (writableProperty.Key)
```

Comment:
Could use a comment about what the key is expected to be.

Reply:
sure, I'll add